### ISSUE TITLE
Return value of `f(x: Sequence[Union[float, T]]) -> C[T]` has wrong T inferred

### ISSUE THREAD

**Issue by rshin** [microsoft/pyright#2962]
**Describe the bug**
Pyright incorrectly infers the return type of a call of a function like `f(x: Sequence[Union[float, T]]) -> C[T]:`. See below for a minimal example.

`create_container(arg)` on the last line should have type `Container[int]`, but it is inferred as `Container[float]`.

**To Reproduce**
See code below.

**Expected behavior**
The code below should not have any type-checking errors.

**Screenshots or Code**
```python
from dataclasses import dataclass
from typing import Generic, Sequence, Union, TypeVar

T = TypeVar("T")

@dataclass
class Container(Generic[T]):
    values: Sequence[Union[float, T]]

def create_container(values: Sequence[Union[float, T]]) -> Container[T]:
    return Container(values)

arg: Sequence[Union[float, int]] = (1, 2.0)
x: Container[int] = create_container(arg)
```

The above code is in Python 3.7.

**VS Code extension or command-line**
Pylance v2022.1.5 in VS Code.

**Comment by erictraut** [comment]
Any time you combine a type variable in a union with other concrete types, you create ambiguous situations for the TypeVar constraint solver because there can be multiple valid solutions. Pyright has a bunch of heuristics in place to "guess" the most desirable solution, but sometimes these heuristics fail. I'll see if I can adjust the heuristics to make this case work without breaking any of the other cases that the existing heuristics are meant to cover.

**Comment by erictraut** [comment]
As I feared, changing this heuristic to work for this case breaks at least one other case. However, I think there's a good argument that this case should be prioritized. Mypy doesn't work in the other case either. So I've decided to break the other case in favor of this one. This change will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.217, which I just published. It will also be included in the next release of pylance.
